go/pio/cmd/pin-perf/main.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2016 Marc-Antoine Ruel. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+// pin-perf does a simple performance test of gpio via sysfs vs. gpio using memory mapped regs
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/maruel/dlibox/go/pio/host"
+	"github.com/maruel/dlibox/go/pio/host/sysfs"
+	"github.com/maruel/dlibox/go/pio/protocols/gpio"
+)
+
+const (
+	// Sleep for a short delay as there can be some capacitance on the line,
+	// requiring a few CPU cycles before the input stabilizes to the new value.
+	shortDelay = time.Nanosecond
+
+	// Purely to help diagnose issues.
+	longDelay = 2 * time.Second
+)
+
+// getPin looks up the pin by number and returns it's memory mapped and sysfs version
+func getPin(s string) (gpio.PinIO, gpio.PinIO, error) {
+	number, err := strconv.Atoi(s)
+	if err != nil {
+		return nil, nil, err
+	}
+	var pm, ps gpio.PinIO
+	ps, err = sysfs.PinByNumber(number)
+	if err != nil {
+		return nil, nil, err
+	}
+	pm = gpio.ByNumber(number)
+	if ps == nil {
+		return nil, nil, errors.New("invalid pin number")
+	}
+	return pm, ps, nil
+}
+
+func testOutput(p gpio.PinIO) (float64, error) {
+	if err := p.Out(gpio.High); err != nil {
+		return 0, err
+	}
+	for i := 0; i < 10; i++ {
+		p.Out(gpio.Level(i&1 == 0))
+	}
+	t0 := time.Now()
+	for i := 0; i < 10000; i++ {
+		p.Out(gpio.Level(i&1 == 0))
+	}
+	return time.Since(t0).Seconds() / 10000, nil
+}
+
+func testInput(p gpio.PinIO) (float64, error) {
+	if err := p.In(gpio.Float); err != nil {
+		return 0, err
+	}
+	for i := 0; i < 10; i++ {
+		p.Read()
+	}
+	t0 := time.Now()
+	for i := 0; i < 10000; i++ {
+		p.Read()
+	}
+	return time.Since(t0).Seconds() / 10000, nil
+}
+
+func testToggle(p1, p2 gpio.PinIO) (float64, error) {
+	if err := p2.In(gpio.Float); err != nil {
+		return 0, err
+	}
+	if err := p1.Out(gpio.Low); err != nil {
+		return 0, err
+	}
+	for i := 0; i < 10; i++ {
+		o := gpio.Level(i&1 == 0)
+		p1.Out(o)
+		in := p2.Read()
+		if in != o {
+			return 0, fmt.Errorf("output %v but read %v", p1, p2)
+		}
+	}
+	t0 := time.Now()
```

Comment:
I would have liked to use test.Benchmark but as far as I remember the code is not exported.